File: apps/admin/src/test/handlers/scheduler.ts
import { http, HttpResponse } from 'msw'
import { NOW, HOUR, DAY } from './shared'

export const mockSchedulerJobs = [
  {
    id: 'job-1',
    name: 'Daily Digest',
    description: 'Generate and post daily summary of Jira and Confluence updates to Slack',
    cronExpression: '0 9 * * *',
    timezone: 'Asia/Seoul',
    jobType: 'AGENT' as const,
    mcpServerName: null,
    toolName: null,
    toolArguments: {},
    agentPrompt: 'Summarize all Jira ticket updates and new Confluence pages from the last 24 hours. Format as a digest and post to the #daily-updates Slack channel.',
    personaId: 'persona-1',
    agentSystemPrompt: null,
    agentModel: 'claude-sonnet-4-20250514',
    agentMaxToolCalls: 20,
    slackChannelId: 'C0123DAILY',
    teamsWebhookUrl: null,
    retryOnFailure: true,
    maxRetryCount: 2,
    executionTimeoutMs: 120000,
    enabled: true,
    lastRunAt: NOW - 14 * HOUR,
    lastStatus: 'SUCCESS' as const,
    lastResult: 'Generated digest with 12 items',
    lastResultPreview: 'Generated digest with 12 items from Jira and Confluence',
    lastFailureReason: null,
    createdAt: NOW - 14 * DAY,
    updatedAt: NOW - 14 * HOUR,
  },
  {
    id: 'job-2',
    name: 'Swagger Sync',
    description: 'Sync all registered Swagger/OpenAPI spec sources',
    cronExpression: '0 */6 * * *',
    timezone: 'UTC',
    jobType: 'MCP_TOOL' as const,
    mcpServerName: 'swagger',
    toolName: 'swagger_sync_all',
    toolArguments: { includeArchived: false },
    agentPrompt: null,
    personaId: null,
    agentSystemPrompt: null,
    agentModel: null,
    agentMaxToolCalls: null,
    slackChannelId: null,
    teamsWebhookUrl: null,
    retryOnFailure: true,
    maxRetryCount: 3,
    executionTimeoutMs: 60000,
    enabled: true,
    lastRunAt: NOW - 26 * HOUR,
    lastStatus: 'SUCCESS' as const,
    lastResult: 'Synced 3 sources',
    lastResultPreview: 'Synced 3 spec sources, 0 changes detected',
    lastFailureReason: null,
    createdAt: NOW - 20 * DAY,
    updatedAt: NOW - 26 * HOUR,
  },
  {
    id: 'job-3',
    name: 'Weekly Report',
    description: 'Compile weekly metrics report and email to stakeholders',
    cronExpression: '0 8 * * 1',
    timezone: 'Asia/Seoul',
    jobType: 'AGENT' as const,
    mcpServerName: null,
    toolName: null,
    toolArguments: {},
    agentPrompt: 'Compile a weekly report of key platform metrics including API usage, tool call counts, feedback scores, and cost analysis. Send to ops@example.com.',
    personaId: null,
    agentSystemPrompt: 'You are an operations analyst assistant.',
    agentModel: 'claude-opus-4-20250514',
    agentMaxToolCalls: 30,
    slackChannelId: null,
    teamsWebhookUrl: null,
    retryOnFailure: false,
    maxRetryCount: 0,
    executionTimeoutMs: 300000,
    enabled: false,
    lastRunAt: NOW - 8 * DAY,
    lastStatus: 'FAILED' as const,
    lastResult: null,
    lastResultPreview: null,
    lastFailureReason: 'Email service unavailable',
    createdAt: NOW - 30 * DAY,
    updatedAt: NOW - 8 * DAY,
  },
]

function getMockSchedulerJobs() {
  return [
    {
      ...mockSchedulerJobs[0],
      name: '일일 요약',
      description: 'Jira와 Confluence 업데이트의 일일 요약을 생성하여 Slack에 게시',
      agentPrompt: '지난 24시간 동안의 모든 Jira 티켓 업데이트와 새 Confluence 페이지를 요약하세요. 다이제스트 형식으로 정리하여 #daily-updates Slack 채널에 게시하세요.',
      lastResult: '12개 항목으로 요약 생성 완료',
      lastResultPreview: 'Jira와 Confluence에서 12개 항목으로 요약 생성 완료',
    },
    {
      ...mockSchedulerJobs[1],
      name: 'Swagger 동기화',
      description: '등록된 모든 Swagger/OpenAPI 스펙 소스 동기화',
      lastResult: '3개 소스 동기화 완료',
      lastResultPreview: '3개 스펙 소스 동기화 완료, 변경 사항 없음',
    },
    {
      ...mockSchedulerJobs[2],
      name: '주간 보고서',
      description: '주간 메트릭 보고서를 작성하여 이해관계자에게 이메일 전송',
      agentPrompt: 'API 사용량, 도구 호출 횟수, 피드백 점수, 비용 분석을 포함한 주간 주요 플랫폼 메트릭 보고서를 작성하세요. ops@example.com로 전송하세요.',
      agentSystemPrompt: '당신은 운영 분석 어시스턴트입니다.',
      lastFailureReason: '이메일 서비스 사용 불가',
    },
  ]
}

export const mockSchedulerExecutions = [
  {
    id: 'exec-1',
    jobId: 'job-1',
    jobName: 'Daily Digest',
    status: 'SUCCESS' as const,
    result: 'Generated digest with 12 items from Jira and Confluence. Posted to #daily-updates.',
    resultPreview: 'Generated digest with 12 items from Jira and Confluence',
    failureReason: null,
    durationMs: 15400,
    dryRun: false,
    startedAt: NOW - 14 * HOUR,
    completedAt: NOW - 14 * HOUR + 15400,
  },
  {
    id: 'exec-2',
    jobId: 'job-1',
    jobName: 'Daily Digest',
    status: 'SUCCESS' as const,
    result: 'Generated digest with 8 items.',
    resultPreview: 'Generated digest with 8 items',
    failureReason: null,
    durationMs: 12300,
    dryRun: false,
    startedAt: NOW - 38 * HOUR,
    completedAt: NOW - 38 * HOUR + 12300,
  },
]

function getMockSchedulerExecutions() {
  return [
    {
      ...mockSchedulerExecutions[0],
      jobName: '일일 요약',
      result: 'Jira와 Confluence에서 12개 항목으로 요약 생성 완료. #daily-updates에 게시됨.',
      resultPreview: 'Jira와 Confluence에서 12개 항목으로 요약 생성 완료',
    },
    {
      ...mockSchedulerExecutions[1],
      jobName: '일일 요약',
      result: '8개 항목으로 요약 생성 완료.',
      resultPreview: '8개 항목으로 요약 생성 완료',
    },
  ]
}

export const schedulerHandlers = [
  http.get('/api/scheduler/jobs', () => {
    return HttpResponse.json(getMockSchedulerJobs())
  }),

  http.get('/api/scheduler/jobs/:id', ({ params }) => {
    const jobs = getMockSchedulerJobs()
    const job = jobs.find(j => j.id === params.id)
    if (!job) return HttpResponse.json({ error: 'Not found' }, { status: 404 })
    return HttpResponse.json(job)
  }),

  http.post('/api/scheduler/jobs', async ({ request }) => {
    const body = await request.json() as Record<string, unknown>
    return HttpResponse.json(
      { id: 'job-new', ...body, enabled: true, lastRunAt: null, lastStatus: null, lastResult: null, lastResultPreview: null, lastFailureReason: null, createdAt: NOW, updatedAt: NOW },
      { status: 201 },
    )
  }),

  http.put('/api/scheduler/jobs/:id', async ({ params, request }) => {
    const body = await request.json() as Record<string, unknown>
    const jobs = getMockSchedulerJobs()
    const job = jobs.find(j => j.id === params.id)
    if (!job) return HttpResponse.json({ error: 'Not found' }, { status: 404 })
    return HttpResponse.json({ ...job, ...body, updatedAt: NOW })
  }),

  http.delete('/api/scheduler/jobs/:id', () => {
    return new HttpResponse(null, { status: 204 })
  }),

  http.post('/api/scheduler/jobs/:id/trigger', () => {
    return HttpResponse.json('작업이 성공적으로 트리거되었습니다')
  }),

  http.post('/api/scheduler/jobs/:id/dry-run', () => {
    return HttpResponse.json('드라이런 완료: 현재 설정으로 실행됩니다')
  }),

  http.get('/api/scheduler/jobs/:id/executions', () => {
    return HttpResponse.json(getMockSchedulerExecutions())
  }),
]

File: src/reactor/api/schemas/scheduler.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ScheduledJobRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    cron_expression: str = Field(alias="cronExpression", min_length=1)
    timezone: str = "Asia/Seoul"
    job_type: str = Field(default="MCP_TOOL", alias="jobType")
    mcp_server_name: str | None = Field(default=None, alias="mcpServerName")
    tool_name: str | None = Field(default=None, alias="toolName")
    tool_arguments: dict[str, Any] = Field(default_factory=dict, alias="toolArguments")
    agent_prompt: str | None = Field(default=None, alias="agentPrompt")
    persona_id: str | None = Field(default=None, alias="personaId")
    agent_system_prompt: str | None = Field(default=None, alias="agentSystemPrompt")
    agent_model: str | None = Field(default=None, alias="agentModel")
    agent_max_tool_calls: int | None = Field(default=None, alias="agentMaxToolCalls", ge=0)
    tags: set[str] = Field(default_factory=set)
    slack_channel_id: str | None = Field(default=None, alias="slackChannelId")
    teams_webhook_url: str | None = Field(default=None, alias="teamsWebhookUrl")
    retry_on_failure: bool = Field(default=False, alias="retryOnFailure")
    max_retry_count: int = Field(default=3, alias="maxRetryCount", ge=0)
    execution_timeout_ms: int | None = Field(default=None, alias="executionTimeoutMs", ge=1000)
    enabled: bool = True


class ScheduledJobResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    name: str
    description: str | None
    cron_expression: str = Field(alias="cronExpression")
    timezone: str
    job_type: str = Field(alias="jobType")
    mcp_server_name: str | None = Field(alias="mcpServerName")
    tool_name: str | None = Field(alias="toolName")
    tool_arguments: dict[str, Any] = Field(alias="toolArguments")
    agent_prompt: str | None = Field(alias="agentPrompt")
    persona_id: str | None = Field(alias="personaId")
    agent_system_prompt: str | None = Field(alias="agentSystemPrompt")
    agent_model: str | None = Field(alias="agentModel")
    agent_max_tool_calls: int | None = Field(alias="agentMaxToolCalls")
    tags: set[str]
    slack_channel_id: str | None = Field(alias="slackChannelId")
    teams_webhook_url: str | None = Field(alias="teamsWebhookUrl")
    retry_on_failure: bool = Field(alias="retryOnFailure")
    max_retry_count: int = Field(alias="maxRetryCount")
    execution_timeout_ms: int | None = Field(alias="executionTimeoutMs")
    enabled: bool
    last_run_at: int | None = Field(alias="lastRunAt")
    last_status: str | None = Field(alias="lastStatus")
    last_result: str | None = Field(alias="lastResult")
    last_result_preview: str | None = Field(alias="lastResultPreview")
    last_failure_reason: str | None = Field(alias="lastFailureReason")
    created_at: int = Field(alias="createdAt")
    updated_at: int = Field(alias="updatedAt")


class ScheduledJobExecutionResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    job_id: str = Field(alias="jobId")
    job_name: str = Field(alias="jobName")
    status: str
    result: str | None
    result_preview: str | None = Field(alias="resultPreview")
    failure_reason: str | None = Field(alias="failureReason")
    duration_ms: int = Field(alias="durationMs")
    dry_run: bool = Field(alias="dryRun")
    started_at: int = Field(alias="startedAt")
    completed_at: int | None = Field(alias="completedAt")


class PaginatedScheduledJobsResponse(BaseModel):
    items: list[ScheduledJobResponse]
    total: int
    offset: int
    limit: int


class PaginatedScheduledExecutionsResponse(BaseModel):
    items: list[ScheduledJobExecutionResponse]
    total: int
    offset: int
    limit: int


class SchedulerTriggerResponse(BaseModel):
    result: str
    dry_run: bool | None = Field(default=None, alias="dryRun")
